src/Confluent.Kafka/Producer.cs

```diff
@@ -30,115 +30,233 @@
     ///     The serialized value.
     /// </returns>
-    public delegate byte[] Serializer<T>(string topic, T data);
+    public delegate byte[] Serializer<T>(T data);
 
-    /// <summary>
-    ///     A generator for <see cref="Confluent.Kafka.Serializer{T}" /> instances.
-    /// </summary>
-    /// <param name="forKey">
-    ///     Whether or not the serializer is for use with keys.
-    /// </param>
-    /// <returns>
-    ///     The serializer.
-    /// </returns>
-    public delegate Serializer<T> SerializerGenerator<T>(bool forKey);
 
     /// <summary>
-    ///     Implements a high-level Apache Kafka producer with key
-    ///     and value serialization.
+    ///     A high-level Apache Kafka producer (without serialization).
     /// </summary>
-    public class Producer<TKey, TValue> : IProducer<TKey, TValue>
-    {
-        private readonly Producer ownedClient;
-        private readonly Handle handle;
-        private Producer producer;
-        private Serializer<TKey> keySerializer;
-        private Serializer<TValue> valueSerializer;
-
-        private void setAndValidateSerializers(Serializer<TKey> keySerializer, Serializer<TValue> valueSerializer)
+    public class Producer : IProducer
+    {        
+        private bool disposeHasBeenCalled = false;
+        private object disposeHasBeenCalledLockObj = new object();
+
+        private readonly bool manualPoll = false;
+        internal readonly bool enableDeliveryReports = true;
+        internal readonly bool enableDeliveryReportKey = true;
+        internal readonly bool enableDeliveryReportValue = true;
+        internal readonly bool enableDeliveryReportTimestamp = true;
+        internal readonly bool enableDeliveryReportHeaders = true;
+
+        private Dictionary<Type, object> serializers = new Dictionary<Type, object>();
+
+        private readonly SafeKafkaHandle kafkaHandle;
+
+        private readonly Task callbackTask;
+        private readonly CancellationTokenSource callbackCts;
+
+        private const int POLL_TIMEOUT_MS = 100;
+        private Task StartPollTask(CancellationToken ct)
+            => Task.Factory.StartNew(() =>
+                {
+                    try
+                    {
+                        while (true)
+                        {
+                            ct.ThrowIfCancellationRequested();
+                            kafkaHandle.Poll((IntPtr)POLL_TIMEOUT_MS);
+                        }
+                    }
+                    catch (OperationCanceledException) {}
+                }, ct, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+
+        private readonly Librdkafka.ErrorDelegate errorCallbackDelegate;
+        private void ErrorCallback(IntPtr rk, ErrorCode err, string reason, IntPtr opaque)
+        {
+            // Ensure registered handlers are never called as a side-effect of Dispose/Finalize (prevents deadlocks in common scenarios).
+            if (kafkaHandle.IsClosed) { return; }
```

Comment:
I don't know and I'm worried about inadvertently introducing a deadlock scenario which would be worse if I did. I've made a note of this to come back to as it could take a fair bit of time and there are more pressing things.